Initialize _fisher_plugins using fish_plugins file (#741)

charego opened the pull request:
This change is for users who commit fish_plugins and plugin sources but omit fish_variables.

On a system where Fisher's universal variables are unset, most Fisher commands do not work out-of-the-box:

* `fisher list [<regex>]`
  * returns nothing
* `fisher remove repo/plugin`
  * fails with error: "Plugin not installed: repo/plugin"
* `fisher update repo/plugin`
  * fails with error: "Plugin not installed: repo/plugin"
* `fisher update`
  * fails with error about conflicting files, deletes fish_plugins file
* `fisher install repo/plugin`
  * if fish_plugins contains the plugin, same behavior as bare `fisher update`
  * otherwise, installs the plugin, rewrites fish_plugins file with only "repo/plugin"

As of this commit all Fisher commands work except for `fisher remove`. Fisher still has no way of knowing which files to remove absent the universal variable that tracks the files associated to a plugin. It may make sense to reject calls like `fisher remove repo/plugin` if the `_fisher_repo_2F_plugin_files` universal variable is missing. Fisher could suggest the user run `fisher update` and try again.

Note: I haven't really investigated how this works with local plugins since I don't use those.

Attempt to fix https://github.com/jorgebucaran/fisher/issues/741.

charego commented:
Bump for further consideration, @jorgebucaran. Thanks.

jorgebucaran commented:
Thanks, @charego. I'll have a look at this as soon as I can manage. 👍

jorgebucaran commented:
I have been addressing and investigating issues all weekend, so I am taking a break for now. However, I wanted to let you know that I will certainly look into this matter soon and will incorporate this change (or a similar one, as I'm not certain this will be merged as is) before Fisher 5. This should occur in the not-too-distant future. 🙏

charego committed:
Initialize _fisher_plugins using fish_plugins file (#741)

This change is for users who commit fish_plugins and plugin sources but
omit fish_variables. On a system where Fisher's universal variables are
unset, most Fisher commands were not working out-of-the-box:

* `fisher install <plugin>` - installs <plugin>, rewrites fish_plugins file with only <plugin>
                            - if <plugin> is in fish_plugins, same behavior as `fisher update`
* `fisher remove  <plugin>` - fails with error: Plugin not installed "<plugin>"
* `fisher update  <plugin>` - fails with error: Plugin not installed "<plugin>"
* `fisher update`           - fails with error about conflicting files, deletes fish_plugins file
* `fisher list   [<regex>]` - returns nothing

As of this commit all Fisher commands work except for `fisher remove`;
Fisher still has no way of knowing which files to remove absent the
universal variable that tracks the files associated to a plugin.

It may make sense to reject calls like `fisher remove <plugin>` if the
`_fisher_<plugin>_files` universal variable is missing. Fisher could
suggest the user run `fisher update` and try again.

ZuBB commented:
is issue described here possibly related with one I have

```sh
vv@my-precious ~/.config/fish/conf.d $ fisher install oh-my-fish/plugin-bang-bang
fisher install version 4.4.3
Fetching https://api.github.com/repos/oh-my-fish/plugin-bang-bang/tarball/HEAD
fisher: Invalid plugin name or host unavailable: "oh-my-fish/plugin-bang-bang"
vv@my-precious ~/.config/fish/conf.d $ ll
```

jorgebucaran commented:
I doubt it. It might be related to the curl or tar version or your OS. What's your OS, distro, and the curl and tar version you're using?

https://github.com/jorgebucaran/fisher/blob/59228d6215cb4c36aff77008751736f540dc42d8/functions/fisher.fish#L101-L106

ZuBB commented:
> I doubt it. It might be related to the curl or tar version or your OS. What's your OS, distro, and the curl and tar version you're using?

its latest macos with m1 chip

```sh
vv@my-precious ~/work/own/careerpartner/src $ curl -V
curl 7.88.1 (x86_64-apple-darwin22.0) libcurl/7.88.1 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.11 nghttp2/1.51.0
Release-Date: 2023-02-20
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe UnixSockets
vv@my-precious ~/work/own/careerpartner/src $ tar --version
bsdtar 3.5.3 - libarchive 3.5.3 zlib/1.2.11 liblzma/5.0.5 bz2lib/1.0.8 
vv@my-precious ~/work/own/careerpartner/src $ 
```

ZuBB commented:
okay, I figured it out 

https://github.com/jorgebucaran/fisher/pull/771

charego commented:
Essentially, how to use fisher with dotfiles? Is there a best practice as you see it? The instructions are geared toward interactive use.

jorgebucaran commented:
Honestly, the reason I haven't been able to work on this is because I don't understand the aim at a fundamental level. Do you just want Fisher to fail silently if universal variables are not set?

itsfarseen commented:
@jorgebucaran, from what I understand, the idea is to not use `fish_variables` to store fisher state as some of us don't commit `fish_variables` to our dotfiles repo.

So our options are:
* Reconstruct the state from the `fish_plugins` file and the list of installed files in `~/.config/fish/`;
* Provide a flag to `fisher update` that would let us overwrite the files present in `~/.config/fish`, but absent in `fish_variables`.

itsfarseen commented:
@jorgebucaran One more thing to consider before merging this, maybe we _should_ commit `fish_variables` to the dotfiles repo.

Plugins _should not_ store non-portable data in `fish_variables`.

For reference:
https://github.com/jorgebucaran/fisher/issues/611
https://github.com/jethrokuan/z/issues/104

jorgebucaran commented:
> @itsfarseen from what I understand, the idea is to not use fish_variables to store fisher state as some of us don't commit fish_variables to our dotfiles repo.

That sounds a lot like https://github.com/jorgebucaran/fisher/issues/611. Fisher preserves the '~' in paths when saving plugins to the universal state, making it safer to commit fish_variables. While it might not be perfect, it does prevent committing data that obviously varies across machines, such as different usernames or home directories. What else are we missing?

itsfarseen commented:
@jorgebucaran

> Fisher preserves the '~' in paths when saving plugins to the universal state

Yes, but Z doesn't. And possibly other plugins too.

OP's solution to this is to not commit `fish_variables` to the dotfiles repo.

My opinion is that, ideally we should ask Z and other problematic plugins to implement proper handling of `fish_variables`, instead of trying to make fisher work with missing `fish_variables`.

VinzNL commented:
Fish itself currently doesn't want you to share the fish_variables file across machines. For example, I share my dotfiles between two machines, one has fish 3.7.1 and one has fish 4.0.0. The first one has SETUVAR __fish_initialized:3400 and the second one has SETUVAR __fish_initialized:3800. There's even discussion about removing (default) uvars altogether, but that doesn't seem to be moving forwards quickly (fish-shell/fish-shell#7317 and fish-shell/fish-shell#7379). So we really need a way to sync plugins across machines without sharing fish_variables for the foreseeable future.
